Add TypeScript definition

Codex- opened the pull request:
Removed my request from DefinitelyTyped and migrated following your [guide.](https://github.com/sindresorhus/typescript-definition-style-guide)

Hopefully this meets the requirements :)

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,92 @@
+ * @returns a Promise with a unhook() method which, when called, unhooks
+ * the streams and resolves the Promise.
+ */
+export function stderr(opts: Options, transform: Transform): PromiseUnhook;
```
`opts` => `options`

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,92 @@
+ * Hooks stderr or alternatively the streams in options.
+ * 
+ * @param opts
+ * @param transform - the transform function.
```
Sentences should start with an uppercase character.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,92 @@
+ * Promise with a unhook() method which, when called, resolves the Promise
+ * with an empty result.
+ */
+export interface PromiseUnhook extends Promise<any> {
```
I would rather call this `HookPromise` as that's what it is. It's a Promise for the hook.

sindresorhus replied:
`any` => `void`?

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,92 @@
+import {Writable} from 'stream';
+
+/**
+ * unhook() method which, when called, unhooks from a stream
```
unhook() should be wrapped in backticks.

Codex- replied:
Ah I see, wasn't sure if this was supported, cheers

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,92 @@
+ * second argument.
+ *
+ * Optionally, when in silent mode, you may return a boolean to influence
+ * the return value of .write(...).
```
.write(...) should be wrapped in backticks.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,92 @@
+export function stdout(transform: Transform): PromiseUnhook;
+
+/**
+ * Hooks stdout or alternatively the streams in options.
```
> alternatively the streams in options.

This is not correct. You're not supposed to use the `streams` option with the `.stdout`/`stderr` methods. I guess that needs to be made clearer in the docs.

Codex- replied:
Hmmm I suppose that means I should remove these:
`stderr(options: Options, transform: Transform): HookPromise;`
`stdout(options: Options, transform: Transform): HookPromise;`

Would you prefer a second set of options be typed in such a way that doesn't allow options with streams as parameter types for `stdout` and `stderr`?

I misunderstood originally but I think I understand that if you use hookstd via `var hookStd = require('hook-std')` then the call `hookStd(opts, str => {})` would be what consumes the stream options, so will need to rework some of the typings to suit this

sindresorhus replied:
You could do something like this for just the default export:

```ts
type DefaultOptions = Options & {
	streams?: Writable[];
}

export default function hookStd(options?: DefaultOptions, transform: Transform): PromiseUnhook;
```

And remove `streams` from the `Options` interface

Codex- replied:
I created a `DefaultOptions` interface which extends the `Options` interface to create this typing.

Thanks for the suggestion, have implemented similar, however you can't follow an optional param with a required one.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,92 @@
+ * Hooks stdout or alternatively the streams in options.
+ * 
+ * @param opts
+ * @param transform - the transform function.
```
> If the parameter description just repeats the parameter name, leave it out.

Codex- replied:
Can you please clarify if this means to remove the whole `@param` doc or just the description?

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,92 @@
+
+/**
+ * Receives stream output as the first argument and the unhook method as the
+ * second argument.
```
Don't hard-wrap. Applies to all the comments.

sindresorhus commented:
Let me know if there's anything in the style guide that could be clearer or improved. It's still a work-in-progress :)

Codex- commented:
Re: the guide. I suppose the main thing I found a little tricky at first was how to use the type assertions correctly, I ended up going through a few other projects to understand the usage more correctly.

sindresorhus commented:
Also note:

> For packages with a default export, use `export default foo;`, not `export = foo;`. You will have to add `module.exports.default = foo;` to the `index.js` file.

sindresorhus commented:
You also left out types for the default export.

Codex- commented:
Yup I've come to realize this after the comment around the stream options, will sort out soonish :)

Codex- commented:
I've _hopefully_ this time covered everything :)

Codex- commented:
Mistakes were made...

Codex- commented:
Should be good now.

Codex- commented:
Actually I take it back, few issues to look at still, will comment when ready for another review

Codex- commented:
Updated with the default export and fixed some typings. I think this is more up to scratch now. Sorry for the delays, uni has been crazy busy.

Can you please clarify a couple of things:

`If the parameter description just repeats the parameter name, leave it out.`
Does this mean I should remove all docstring params that are simply `@param transform` or that there shouldn't be a description if the description just repeats the param such as `@param transform the transform function`.


In the [readme](https://github.com/sindresorhus/hook-std#hookstdoptions-transform) it specifies `unhooks both stdout and stderr` but if you use custom or additional streams it unhooks these too, should it in that case read something more along the lines of `unhooks both stdout, stderr or any specified streams`

Cheers

sindresorhus commented:
> Does this mean I should remove all docstring params that are simply @param transform

Yes

> In the readme it specifies unhooks both stdout and stderr but if you use custom or additional streams it unhooks these too, should it in that case read something more along the lines of unhooks both stdout, stderr or any specified streams

Maybe just:

> Unhooks the streams in the `streams` option.

sindresorhus commented on index.d.ts:
```diff
@@ -0,0 +1,96 @@
+ *
+ * @returns A `Promise` with a `unhook()` method which, when called, unhooks the streams and resolves the `Promise`.
+ */
+export function stdout(options: Options, transform: Transform): HookPromise;
```
Why make this an overload instead of just marking `options` as optional?

Codex- replied:
Optional parameters cannot precede required parameters.

sindresorhus replied:
Hmm, ok. That's a lame limitation of TS...

Codex- commented:
Just wondering if there were any pending changes you were wanting that I have possibly missed?

sindresorhus commented:
Thank you :)

Codex- commented:
Great, thanks Sindre!
